File: add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        head = ListNode(0)
        tail = head
        carry = 0
        while l1 != None or l2 != None or carry != 0:
            dig1 = l1.val if l1 != None else 0
            dig2 = l2.val if l2 != None else 0

            sum = dig1 + dig2 + carry
            carry = sum // 10
            dig3 = sum % 10

            tail.next = ListNode(dig3)
            tail = tail.next

            l1 = l1.next if l1 != None else None
            l2 = l2.next if l2 != None else None
        
        ans = head.next
        head = None
        return ans
        

File: add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
    public:
        ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
            int digCount{1};
            ListNode* head = new ListNode();
            ListNode * tail = head;
            int carry{};
    
            while (l1 != nullptr || l2 != nullptr || carry != 0) {
    
                int dig1 = (l1 != nullptr) ? l1->val : 0;
                int dig2 = (l2 != nullptr) ? l2->val : 0;
                
                int sum = dig1 + dig2 + carry;
                int dig3 = sum % 10;
                carry = sum / 10;
    
                ListNode* ans = new ListNode(dig3);
                tail->next = ans;
                tail = tail->next;
                
                l1 = (l1 != nullptr) ? l1->next : nullptr;
                l2 = (l2 != nullptr) ? l2->next : nullptr;
            }
            head = head->next;
            return head;
    
    
        }
    };